### ISSUE TITLE
[Question] Unable to finetune English Quartznet 15x5 with sentencepieces decoder.

### ISSUE THREAD

**Issue by jonaskratochvil** [NVIDIA/NeMo#1269]
Hello,

I have been using Quartznet 15x5 model for some time and recently I have tried to use the sentencepieces model trained on my data to encode the text and finetune the network with this setup. Unfortunately my model always diverges and eventually ends-up predicting only the blank symbol all the time. I have checked the correctness of my encoding multiple times and everything seems to be correct. Below I am listing experiments that I have tried:

1. Finetune on large train set with 500 sentencepieces (diverged)
2. Overfit 50 examples with 50 sentencepieces (diverged)
3. Overfit decoder from scratch on character sentencepieces (converged)
4. Try finetuning with freezing whole encoder for 100 epochs and finetuning only decoder (diverged for sentencepieces, converged for characters)
5. Use multiple learning rate setups with sentencepieces (diverged)
6. Use only unigram and bigram sentencepieces (diverged)

Have you managed to get any sensible results with sentencepieces encoding? Would you have any tips you could share regarding the training process? I am especially confused that my model is not even able to overfit to small number of examples, this would suggest bug somewhere in my encoding pipeline but when I use the same setup and use only character level sentencepieces I am able to overfit the model. Any advice would be appreciated.

Jonas

**Comment by okuchaiev** [comment]
QuartzNet is a character-based model. I would recommend you try 'ContextNet-192-WPE-1024-8x-Stride' pre-trained model, here is an example on how to instantiate it:
```
asr_model = nemo_asr.models.EncDecCTCModelBPE.from_pretrained(model_name="ContextNet-192-WPE-1024-8x-Stride")
```
Note that here ``EncDecCTCModelBPE`` is used. This kinds of models (e.g. BPE/sentencepeice based) are work in progress but the above checkpoint is pretty good for En. 

What language are you trying to fine-tune to?  @titu1994  FYI

**Comment by titu1994** [comment]
Just a note about the name above, it's not true ComtextNet (ie with RNNT), but ContextNet CTC. We will be renaming it in the future. 

May I ask whether you were using Nemo 0.11 or 1.0.0 beta release for the above experiments? Nemo 0.11x does not support Sentencepiece decoding during evaluation. 

As to simply computing loss and training, BPE models on general will reduce the target label length by around 2.5-3x. When you say diverged, did the loss become Nan? If so, there is a flag in new Nemo (set by default) called zero_inf. This resolved nan loss cases when alignment is not possible due to CTC limitations. 

Finally, due to the target sequence being 3x smaller, using the same kernel sizes as QuartzNet will make training unstable (especially if stride > 2). Please refer to how we balance the kernels for BPE for ContextNet CTC models in the examples experimental directory.

**Comment by jonaskratochvil** [comment]
Thank you for your answers, 
I am trying to finetune my model for English so the suggested  'ContextNet-192-WPE-1024-8x-Stride'  sound like a good starting point. 
I am using the 1.0.0 beta Nemo version for my experiments. 
When I say diverged I mean that the model eventually learns to predict only blank symbols and is not even able to overfit small number of examples (I have just finished experiment trying to overfit Quartznet 15x5 without and checkpoint and the same divergence behaviour persist). 
From what I can conclude the issue indeed seems to be in the small kernel size, I will try to explore in this direction using the ContexNet.
